File: ru/codes/kotlin/chapter_backtracking/preorder_traversal_iii_compact.kt
/**
 * File: preorder_traversal_iii_compact.kt
 * Created Time: 2024-01-25
 * Author: curtishd (1023632660@qq.com)
 */

package chapter_backtracking.preorder_traversal_iii_compact

import utils.TreeNode
import utils.printTree

var path: MutableList<TreeNode>? = null
var res: MutableList<MutableList<TreeNode>>? = null

/* Прямой обход: пример 3 */
fun preOrder(root: TreeNode?) {
    // Отсечение
    if (root == null || root._val == 3) {
        return
    }
    // Попытаться
    path!!.add(root)
    if (root._val == 7) {
        // Записать решение
        res!!.add(path!!.toMutableList())
    }
    preOrder(root.left)
    preOrder(root.right)
    // Откат
    path!!.removeAt(path!!.size - 1)
}

/* Driver Code */
fun main() {
    val root = TreeNode.listToTree(mutableListOf(1, 7, 3, 4, 5, 6, 7))
    println("\nИнициализировать двоичное дерево")
    printTree(root)

    // Прямой обход
    path = mutableListOf()
    res = mutableListOf()
    preOrder(root)

    println("\nВывести все пути от корня до узла 7, путь не должен содержать узлы со значением 3")
    for (path in res!!) {
        val _vals = mutableListOf<Int>()
        for (node in path) {
            _vals.add(node._val)
        }
        println(_vals)
    }
}

File: ru/codes/swift/chapter_backtracking/preorder_traversal_iii_compact.swift
/**
 * File: preorder_traversal_iii_compact.swift
 * Created Time: 2023-04-30
 * Author: nuomi1 (nuomi1@qq.com)
 */

import utils

var path: [TreeNode] = []
var res: [[TreeNode]] = []

/* Предварительный обход: пример 3 */
func preOrder(root: TreeNode?) {
    // Отсечение
    guard let root = root, root.val != 3 else {
        return
    }
    // Попытка
    path.append(root)
    if root.val == 7 {
        // Записать решение
        res.append(path)
    }
    preOrder(root: root.left)
    preOrder(root: root.right)
    // Откат
    path.removeLast()
}

@main
enum PreorderTraversalIIICompact {
    /* Driver Code */
    static func main() {
        let root = TreeNode.listToTree(arr: [1, 7, 3, 4, 5, 6, 7])
        print("\nИнициализация двоичного дерева")
        PrintUtil.printTree(root: root)

        // Предварительный обход
        path = []
        res = []
        preOrder(root: root)

        print("\nВсе пути от корня к узлу 7, не содержащие узлов со значением 3")
        for path in res {
            var vals: [Int] = []
            for node in path {
                vals.append(node.val)
            }
            print(vals)
        }
    }
}
